lib/Runtime/Executor/ThreadPoolExecutor.cpp

```diff
@@ -0,0 +1,377 @@
+    auto *placeholder = phTensorPair.first;
+    const auto *tensor = phTensorPair.second;
+
+    executionState->insertIntoResultCtx(placeholder, tensor->clone());
+  }
+}
+
+void ThreadPoolExecutor::handleDeviceManagerResult(RunIdentifierTy runId,
+                                                   ResultCode resultCode,
+                                                   std::shared_ptr<Context> ctx,
+                                                   const DAGNode *node) {
+  // Get the execution state for the run.
+  auto it = executionStates_.find(runId);
+
+  if (it == executionStates_.end()) {
+    assert("Execution state should exist!");
+  }
+
+  std::shared_ptr<ExecutionState> executionState = it->second;
+
+  // Set the result code for the run.
+  executionState->setResultCode(resultCode);
+
+  if (resultCode != ResultCode::Executed) {
+    // If the DeviceManager failed to execute the node, call the callback
+    // provided when ThreadPoolExecutor::run() was called, clean up the
+    // lock and execution state for the run and exit.
+    DoneCbTy cb = executionState->getCallback();
+    cb(runId, resultCode, nullptr);
+  } else if (executionState->getResultCode() == ResultCode::Executed) {
+    if ((node->children).empty()) {
+      // If the node has no children, propagate its outputs to the result
+      // Context for the run.
+      propagateOutputPlaceholders(runId, ctx);
+    } else {
+      // If the node has children, propagate its outputs to the input Contexts
+      // for any of its children that need them as inputs.
+      for (auto &child : node->children) {
+        propagatePlaceholdersForNode(runId, child, ctx);
+
+        // Execute any child that has no parent nodes left to execute.
+        bool childReadyToExecute =
+            executionState->incrementNodeParentsDone(child);
+        if (childReadyToExecute) {
+          // Mark the node as "inflight" (i.e. currently executing).
+          executionState->incrementInflightNodes();
+          executeDAGNode(runId, child);
+        }
+      }
+    }
+
+    // Now, check if all nodes in the graph are done. If so, the callback can be
+    // called and all state associated with the run can be erased.
+    bool noNodesInflight = executionState->decrementInflightNodes();
+
+    if (noNodesInflight) {
+      // If there are no nodes inflight, that means all nodes are done. Call
+      // the callback if the entire execution succeeded and erase the state
+      // information. If some node failed, the callback was called in the
+      // invocation of handleDeviceManagerResult() for that node, so the
+      // callback should be skipped here.
+      if (executionState->getResultCode() == ResultCode::Executed) {
+        DoneCbTy cb = executionState->getCallback();
+        cb(runId, executionState->getResultCode(),
+           executionState->getUniqueResultContextPtr());
+      }
+      std::lock_guard<std::mutex> lock(executionStatesMutex_);
```

Comment:
It's weird that we grab the lock down here only? Don't we modify the executionState above? What if two nodes finish at the same time and both start modifying execution state?

Reply:
That cannot happen - `noNodesInFlight` can only be true for one thread.